staging/src/k8s.io/apiserver/pkg/util/flowcontrol/dropped_requests_tracker.go

```diff
@@ -0,0 +1,229 @@
+	// maxRetryAfter represents the maximum possible retryAfter.
+	maxRetryAfter = int64(32)
+)
+
+// DroppedRequestsTracker is an interface that allows tracking
+// a history od dropped requests in the system for the purpose
+// of adjusting RetryAfter header to avoid system overload.
+type DroppedRequestsTracker interface {
+	// RecordDroppedRequest records a request that was just
+	// dropped from processing.
+	RecordDroppedRequest(plName string)
+
+	// GetRetryAfter returns the current suggested value of
+	// RetryAfter value.
+	GetRetryAfter(plName string) int64
+}
+
+// unixStat keeps a statistic how many requests were dropped within
+// a single second.
+type unixStat struct {
+	unixTime int64
+	requests int64
+}
+
+type droppedRequestsStats struct {
+	lock sync.RWMutex
+
+	// history stores the history of dropped requests.
+	history []unixStat
+
+	// To reduce lock-contention, we store the information about
+	// the current second here, which we can then access under
+	// reader lock.
+	currentUnix  int64
+	currentCount atomic.Int64
+
+	retryAfter           atomic.Int64
+	retryAfterUpdateUnix int64
+}
+
+func newDroppedRequestsStats(nowUnix int64) *droppedRequestsStats {
+	result := &droppedRequestsStats{
+		retryAfterUpdateUnix: nowUnix,
+	}
+	result.retryAfter.Store(1)
+	return result
+}
+
+func (s *droppedRequestsStats) recordDroppedRequest(unixTime int64) {
+	// Short path - if the current second matches passed time,
+	// just update the stats.
+	if done := func() bool {
+		s.lock.RLock()
+		defer s.lock.RUnlock()
+		if s.currentUnix == unixTime {
+			s.currentCount.Add(1)
+			return true
+		}
+		return false
+	}(); done {
+		return
+	}
+
+	// We trigger the change of <currentUnix>.
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if s.currentUnix == unixTime {
+		s.currentCount.Add(1)
+		return
+	}
+
+	s.updateHistory(s.currentUnix, s.currentCount.Load())
+	// We only consider decreasing retryAfter when bumping the current second.
+	s.tryDecreasingRetryAfterLocked(s.currentUnix)
+
+	// We only consider increasing retryAfter when bumping the current second.
+	// TODO(wojtek-t): Consider if we don't want to be more reactive.
+	s.tryIncreasingRetryAfterLocked(s.currentUnix)
+
+	s.currentUnix = unixTime
+	s.currentCount.Store(1)
+}
+
+func (s *droppedRequestsStats) updateHistory(unixTime int64, count int64) {
+	s.history = append(s.history, unixStat{unixTime: unixTime, requests: count})
+
+	startIndex := 0
+	for ; startIndex < len(s.history) && unixTime-s.history[startIndex].unixTime > maxRetryAfter; startIndex++ {
+	}
+	if startIndex > 0 {
+		s.history = s.history[startIndex:]
+	}
+}
+
+// tryDecreasingRetryAfterLocked is checking if retryAfter can be decreased and
+// decreases it if decided. It makes the decision if there were less than
+// `retryAfter` dropped requests within the last `retryAfter` seconds.
+func (s *droppedRequestsStats) tryDecreasingRetryAfterLocked(unixTime int64) {
+	retryAfter := s.retryAfter.Load()
+	if retryAfter <= 1 {
```

Comment:
Don't you also want `|| unixTime-s.retryAfterUpdateUnix < retryAfter` condition here too?